### PR TITLE
Allow adding and removing styles at runtime

### LINKED ISSUE #75: Allow styles to be attached/detached.
Currently once a style is attached to a control it is there for good. Need to allow styles to be removed from the tree, which will cause the style to be detached from all attached controls, and attached to the tree.

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

Implements one of the oldest issues still open: #75 - allow styles to be attached/detached. This means that with this PR you should be able to add and remove styles from a `Control.Styles` or `Application.Styles` collection (or any nested style collections within) and the affected controls will be updated with the new or removed styles.

## What is the current behavior?

If a style is added or removed from the application, then existing controls will not notice that a change has happened.

## How was the solution implemented (if it's not obvious)?

**I'm assuming here that adding/removing styles will be an infrequent operation, and optimizing for memory use in the general case rather than making adding/removing styles fast (i.e. we're aiming for zero overhead here).**

When a style is removed from a `Control.Styles` or `Application.Styles` collection then the `Styles` collection calls `IStyleHost.StylesRemoved`, which is implemented on `StyledElement`. This method then calls `DetachStyles` on itself and on all descendant controls. Because a list of applied style instances are stored in each control and the instance has a reference to the style that created it, the style instances that belong to a removed style can be removed.

Similarly when a style is added, `IStyleHost.StylesAdded` is called. Adding a style is a little more complicated because the new style potentially needs to be inserted in between existing styles in the value store of a control, and we don't know exactly where. For this reason, when a style is added _all_ attached styles on the control and its descendants are detached. At the next measure, styling will be reapplied to the entire sub-tree.

## Checklist

- [x] Added unit tests (if possible)?

## Fixed issues

Fixes #75

### CODE REVIEW

File: src/Avalonia.Styling/StyledElement.cs [hunk 1]
```diff
@@ -207,12 +207,13 @@ public Classes Classes
         /// </remarks>
         public Styles Styles
         {
-            get
+            get
```
[the remark this answers is not in the document]
Comment (grokys) [reply src/Avalonia.Styling/StyledElement.cs]: Huh, no idea. I'll revert that.

File: src/Avalonia.Styling/StyledElement.cs [hunk 2]
```diff
@@ -750,5 +832,42 @@ private void ThisResourcesChanged(object sender, ResourcesChangedEventArgs e)
...
+            }
+
+            return result ?? styles;
+        }
+
+        private static void RecurseStyles(IReadOnlyList<IStyle> styles, List<IStyle> result)
+        {
+            var count = styles.Count;
+
+            result.Capacity += count;
```
[the remark this answers is not in the document]
Comment (grokys) [reply src/Avalonia.Styling/StyledElement.cs]: I'll remove this then.

**Review by MarchingCube — APPROVED** [review]
LGTM.

**Comment by waliarubal** [comment]
@grokys Does this work on stable builds? We kind of need this feature to implement theme switching in our app [Jaya](https://github.com/waliarubal/Jaya).

**Comment by grokys** [comment]
@waliarubal no, it's not going to be included in the 0.9.x releases as they only receive bug fixes and this is a reasonably large refactoring. It will be included in 0.10.

**Pull request merged**